File: src/main/java/com/leet/easy/AddBinary.go
package main

/*
67. Add Binary

https://leetcode.com/problems/add-binary/

Easy
Given two binary strings a and b, return their sum as a binary string.

Example 1:
Input: a = "11", b = "1"
Output: "100"

Example 2:
Input: a = "1010", b = "1011"
Output: "10101"

Constraints:
    1 <= a.length, b.length <= 104
    a and b consist only of '0' or '1' characters.
    Each string does not contain leading zeros except for the zero itself.
*/

func addBit(a, b, carry byte) (byte, byte) {
	if a == '0' && b == '0' {
		return carry, '0'
	} else if (a == '1' && b == '0') || (a == '0' && b == '1') {
		if carry == '1' {
			return '0', carry
		} else {
			return '1', carry
		}
	} else {
		return carry, '1'
	}
}

func addBinary(a string, b string) string {
	l1, l2, carry, sum := len(a)-1, len(b)-1, byte('0'), byte('0')
	ans := []byte{}
	for l1 >= 0 && l2 >= 0 {
		sum, carry = addBit(a[l1], b[l2], carry)
		ans = append(ans, sum)
		l1--
		l2--
	}
	for l1 >= 0 {
		sum, carry = addBit(a[l1], '0', carry)
		ans = append(ans, sum)
		l1--
	}
	for l2 >= 0 {
		sum, carry = addBit('0', b[l2], carry)
		ans = append(ans, sum)
		l2--
	}
	if carry == '1' {
		ans = append(ans, carry)
	}
	for i := 0; i < len(ans)/2; i++ {
		ans[i], ans[len(ans)-i-1] = ans[len(ans)-i-1], ans[i]
	}
	return string(ans)
}

File: src/main/java/com/leet/easy/AddBinary.rs
// https://leetcode.com/problems/add-binary/

struct Solution {}

impl Solution {
  pub fn add_binary(a: String, b: String) -> String {
    let (mut ai, mut bi) = (a.len() - 1, b.len() - 1);
    let avec = a.chars().collect::<Vec<char>>();
    let bvec = b.chars().collect::<Vec<char>>();
    let (mut sum, mut carry) = (0_i32, 0_i32);
    let mut ans = vec![];

    while ai < avec.len() || bi < bvec.len() {
      let mut count = carry;
      if ai < avec.len() && ai >= 0 && avec[ai] == '1' {
        count += 1;
      }
      if bi < bvec.len() && bi >= 0 && bvec[bi] == '1' {
        count += 1;
      }
      let res = match count {
        0 => (0, 0),
        1 => (1, 0),
        2 => (0, 1),
        3 => (1, 1),
        _ => (0, 0),
      };
      ai -= 1;
      bi -= 1;
      ans.push(res.0);
      carry = res.1;
    }
    if carry == 1 {
      ans.push(1);
    }
    ans.reverse();
    let rev = ans.iter().map(|e| e.to_string()).collect::<String>();
    rev
  }
}
